prusti-interface/src/specs/mod.rs

```diff
@@ -33,47 +38,189 @@ impl fmt::Debug for SpecItem {
     }
 }
 
-struct Item<'tcx> {
-    name: Symbol,
-    attrs: &'tcx [ast::Attribute],
+struct ProcedureSpecRef {
+    spec_id_refs: Vec<prusti_specs::specifications::common::SpecIdRef>,
+    pure: bool,
+    trusted: bool,
 }
 
+/// Specification collector, intended to be applied as a visitor over the crate
+/// HIR. After the visit, [determine_def_specs] can be used to get back
+/// a mapping of DefIds (which may not be local due to extern specs) to their
+/// [SpecificationSet], i.e. procedures, loop invariants, and structs.
 pub struct SpecCollector<'tcx> {
     tcx: TyCtxt<'tcx>,
+    extern_resolver: ExternSpecResolver<'tcx>,
+
+    /// Collected assertions before deserialisation.
     spec_items: Vec<SpecItem>,
+
     typed_expressions: HashMap<String, LocalDefId>,
-    extern_resolver: ExternSpecResolver<'tcx>,
+
+    /// Collected, deserialised assertions, keyed by their specification id.
+    typed_specs: typed::SpecificationMap<'tcx>,
+
+    /// Resolved specifications.
+    procedure_specs: HashMap<LocalDefId, ProcedureSpecRef>,
+    loop_specs: HashMap<LocalDefId, Vec<SpecificationId>>,
 }
 
 impl<'tcx> SpecCollector<'tcx> {
     pub fn new(tcx: TyCtxt<'tcx>) -> Self {
         Self {
             tcx: tcx,
             spec_items: Vec::new(),
+            typed_specs: HashMap::new(),
+            procedure_specs: HashMap::new(),
+            loop_specs: HashMap::new(),
             typed_expressions: HashMap::new(),
             extern_resolver: ExternSpecResolver::new(tcx),
         }
     }
 
-    pub fn determine_typed_procedure_specs(self) -> typed::SpecificationMap<'tcx> {
-        let typed_expressions = self.typed_expressions;
-        let tcx = self.tcx;
-        self.spec_items
+    fn prepare_typed_procedure_specs(&mut self) {
+        let mut spec_items = self.spec_items.drain(..).collect::<Vec<SpecItem>>();
+        self.typed_specs = spec_items
             .into_iter()
             .map(|spec_item| {
                 let assertion = reconstruct_typed_assertion(
                     spec_item.specification,
-                    &typed_expressions,
-                    tcx
+                    &self.typed_expressions,
+                    self.tcx
                 );
                 (spec_item.spec_id, assertion)
             })
             .collect()
     }
 
-    pub fn determine_extern_procedure_specs(&self, env: &Environment<'tcx>) -> typed::ExternSpecificationMap<'tcx> {
+    pub fn determine_def_specs(&mut self, env: &Environment<'tcx>) -> typed::DefSpecificationMap<'tcx> {
```

Comment:
API nit: As far as I understand `determine_def_specs` is only meant to be called once and, so I would replace the `&mut self` with `self` and I would rename the method to `build_def_specs` to make it clear that it's an instance of the builder pattern.